src/main/java/frc/robot/subsystems/vision/Limelight.java

```diff
@@ -0,0 +1,120 @@
+
+    private final NetworkTable table = NetworkTableInstance.getDefault().getTable("limelight");
+    private final DoubleSubscriber tx = table.getDoubleTopic("tx").subscribe(0.0);
+    private final DoubleSubscriber ty = table.getDoubleTopic("ty").subscribe(0.0);
+    private final IntegerSubscriber tv = table.getIntegerTopic("tv").subscribe(0);
+    private final DoubleSubscriber ts = table.getDoubleTopic("ts").subscribe(0.0);
+    private final IntegerSubscriber tid = table.getIntegerTopic("tid").subscribe(0);
+
+    private Limelight() {
+        super(new LimelightLogInputs());
+        for (int i=5800; i<=5805; i++){
+            PortForwarder.add(i, "limelight.local", i);
+        }
+    }
+
+    /**
+     * If there is no instance of the limelight class it creates one and returns it
+     * @return Limelight instance
+     */
+    public static Limelight getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new Limelight();
+        }
+        return INSTANCE;
+    }
+
+    /**
+     * Checks weather the limelight can see any targets
+     * @return weather the limelight detects any targets
+     */
+    public boolean hasTargets() {
+        return tv.get() == 1;
+    }
+
+    /**
+     * @return aprilTag id
+     */
+    public long getTagId() {
+        return tid.get();
+    }
+
+
+    /**
+     * @return pitch from robot to target
+     */
+    public Rotation2d getPitch() {
+        return Rotation2d.fromDegrees(ty.get());
+    }
+
+    /**
+     * @return distance from target
+     */
+    public OptionalDouble getTargetDistance(double targetHeight) {
+        double totalPitch = Constants.CAMERA_PITCH + getPitch().getRadians();
+        double error = Constants.CAMERA_HEIGHT-targetHeight;
+        if (hasTargets()) {
+            return OptionalDouble.of(Math.abs(error / Math.tan(totalPitch)));
+        }
+        return OptionalDouble.empty();
+    }
+
+    /**
+     * @return robot yaw
+     */
+    public OptionalDouble getYaw() {
+        if (hasTargets()) {
+            return OptionalDouble.of(ts.get());
+        }
+        return OptionalDouble.empty();
+    }
+
+    /**
+     * Estimates the position of the robot
+     * @param robotAngle The angle of the robot
+     * @return Optional Pose2d of the robot coordinates
+     */
+    public Optional<Pose2d> estimatePose(Rotation2d robotAngle, Pose3d target) {
+        double absoluteAngle = robotAngle.getRadians() + getYaw().orElse(0);
+        double xTargetDistance = getTargetDistance(target.getZ()).orElse(0) * Math.sin(absoluteAngle);
+        double yTargetDistance = getTargetDistance(target.getZ()).orElse(0) * Math.cos(absoluteAngle);
+        if (!hasTargets()) {
+            return Optional.empty();
+        }
+        double xDistance = xTargetDistance - target.getX();
+        double yDistance = yTargetDistance - target.getY();
+        Translation2d translation2d = new Translation2d(xDistance, yDistance);
```

Comment:
[`Pose3d` has a pretty extensive API](https://github.wpilib.org/allwpilib/docs/release/java/edu/wpi/first/math/geometry/Pose3d.html). Perhaps it contains a method that makes all this work much easier?